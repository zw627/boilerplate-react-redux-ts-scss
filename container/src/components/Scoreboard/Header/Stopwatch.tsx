import React, { useEffect } from "react";

import { StopwatchProps } from "Containers/Scoreboard/Header/Stopwatch";
import formatTime from "Utils/formatTime";

const Stopwatch: React.FC<StopwatchProps> = ({
  isRunning,
  elapsedTime,
  tick,
  toggle,
  reset,
}: StopwatchProps) => {
  useEffect(() => {
    let intervalId: number | undefined;
    if (isRunning) intervalId = window.setInterval(() => tick(Date.now()), 1);
    else clearInterval(intervalId);
    return (): void => clearInterval(intervalId);
  });

  // Handlers
  function handleToggle(): void {
    toggle(Date.now());
  }
  function hanldeReset(): void {
    reset(Date.now());
  }

  return (
    <div className="stopwatch">
      <h2>Stopwatch</h2>
      <div className="stopwatch-time">{formatTime(elapsedTime)}</div>
      <button onClick={handleToggle}>{isRunning ? "Stop" : "Start"}</button>
      <button onClick={hanldeReset}>Reset</button>
    </div>
  );
};

export default Stopwatch;
