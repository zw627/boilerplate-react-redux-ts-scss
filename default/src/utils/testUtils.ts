// Helper functions for Jest and Enzyme

import { ReactWrapper, ShallowWrapper } from "enzyme";
import { createStore, Store } from "redux";

import { rootReducer } from "Store/index";

// Type for parameters of setupStore()
type setupStorePars = {
  playerList?: {
    id: string;
    name: string;
    score: number;
    created: string;
    updated: string;
    isSelected: boolean;
  }[];
  selectedPlayerId?: string;
  isRunning?: boolean;
  previousTime?: number;
  elapsedTime?: number;
  isLightMode?: boolean;
  visibility?: boolean;
};

// Setup a store with some fake states for testing
export function setupStore({
  playerList = [
    {
      id: "1654A",
      name: "Alpha",
      score: 10,
      created: "11/21/2016",
      updated: "06/12/2018",
      isSelected: false,
    },
    {
      id: "1654B",
      name: "Beta",
      score: 4,
      created: "11/8/2016",
      updated: "11/9/2017",
      isSelected: false,
    },
  ],
  isRunning = false,
  previousTime = 0,
  elapsedTime = 0,
  isLightMode = false,
  visibility = false,
}: setupStorePars = {}): Store {
  return createStore(rootReducer, {
    notification: {
      storage: {
        visibility,
      },
    },
    scoreboard: {
      player: {
        playerList,
      },
      stopwatch: {
        isRunning,
        previousTime,
        elapsedTime,
      },
      theme: {
        isLightMode,
      },
    },
  });
}

// Multiple toEqual, useful when testing a table
export function multiToEqual([...args]: Array<string[] | number[]>): void {
  // [[a, b], [a, b], [a, b]]
  [...args].map((arg) => {
    // [a, b] (expect a equals to b)
    expect(arg[0]).toEqual(arg[1]);
  });
}

export function matchLength(
  wrapper: ReactWrapper | ShallowWrapper,
  [element, number]: [string, number]
): void {
  expect(wrapper.find(element).length).toEqual(number);
}

// Multiple matchLength
export function matchMultiLength(
  wrapper: ReactWrapper | ShallowWrapper,
  [...args]: [string, number][]
): void {
  // [[a, b], [a, b], [a, b]]
  [...args].map((arg) => {
    // [a, b] (expect a equals to b)
    matchLength(wrapper, [arg[0], arg[1]]);
  });
}

// export const matchText = (
//   wrapper: ReactWrapper | ShallowWrapper,
//   [element, index]: [string, number],
//   text: string
// ): void => {
//   expect(
//     wrapper
//       .find(element)
//       .at(index)
//       .text()
//   ).toEqual(text);
// };

// export const matchMultiText = (
//   wrapper: ReactWrapper | ShallowWrapper,
//   [...args]
// ): void => {
//   if ([...args].length > 0)
//     [...args].map(arg => {
//       matchText(wrapper, [arg[0], arg[1]], arg[2]);
//     });
// };
