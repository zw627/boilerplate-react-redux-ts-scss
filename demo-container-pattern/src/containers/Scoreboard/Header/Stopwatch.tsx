import { connect } from "react-redux";

// Component
import Stopwatch from "Components/Scoreboard/Header/Stopwatch";

// Store
import { AppState } from "Store/index";
import { isRunningSelector, elaspedTimeSelector } from "Store/selectors";
import { tick, toggle, reset } from "Store/scoreboard/stopwatch/actions";

interface StopwatchStateProps {
  isRunning: boolean;
  elapsedTime: number;
}

export interface StopwatchProps extends StopwatchStateProps {
  tick: typeof tick;
  toggle: typeof toggle;
  reset: typeof reset;
}

const mapStateToProps = (state: AppState): StopwatchStateProps => ({
  isRunning: isRunningSelector(state),
  elapsedTime: elaspedTimeSelector(state),
});

const actionCreators = {
  tick,
  toggle,
  reset,
};

export default connect(mapStateToProps, actionCreators)(Stopwatch);
