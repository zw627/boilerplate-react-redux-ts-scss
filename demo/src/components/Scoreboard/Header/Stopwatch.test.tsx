import React from "react";
import { Provider } from "react-redux";
import { Store } from "redux";
import { mount, ReactWrapper } from "enzyme";
import mockdate from "mockdate";

import Stopwatch from "Components/Scoreboard/Header/Stopwatch";
import { setupStore } from "Utils/testUtils";

jest.useFakeTimers();
mockdate.set(1574086940640);

function setupWrapper(mockStore = setupStore(), props = {}): ReactWrapper {
  return mount(
    <Provider store={mockStore}>
      <Stopwatch {...props} />
    </Provider>
  ) as ReactWrapper;
}

describe("Scoreboard/Stopwatch", () => {
  let wrapper: ReactWrapper;
  let mockStore: Store;

  // Setup store and wrapper, mock dispatch
  beforeEach(() => {
    mockStore = setupStore();
    mockStore.dispatch = jest.fn();
    wrapper = setupWrapper(mockStore, {});
  });

  test("compare to the last snapshot", () => {
    expect(wrapper.find(Stopwatch)).toMatchSnapshot();
  });

  it('should display time in "m:s.ms" format', () => {
    expect(wrapper.find("div.stopwatch-time").length).toEqual(1);
    expect(wrapper.find("div.stopwatch-time").at(0).text()).toEqual("00:00.00");
  });

  it("should handle tick() dispatch", () => {
    const mockStore = setupStore({ isRunning: true });
    mockStore.dispatch = jest.fn();
    wrapper = setupWrapper(mockStore, {});
    const action = {
      type: "stopwatch/tick",
      payload: {
        date: Date.now(),
      },
    };
    expect(setInterval).toHaveBeenCalledTimes(1);
    jest.advanceTimersByTime(3000);
    expect(mockStore.dispatch).toHaveBeenCalledWith(action);
    expect(mockStore.dispatch).toHaveBeenCalledTimes(3000);
  });

  it("should handle toggle() dispatch", () => {
    const action = {
      type: "stopwatch/toggle",
      payload: {
        date: Date.now(),
      },
    };
    wrapper.find("button").at(0).simulate("click");
    expect(mockStore.dispatch).toHaveBeenCalledWith(action);
    expect(mockStore.dispatch).toHaveBeenCalledTimes(1);
  });

  it("should handle reset() dispatch", () => {
    const action = {
      type: "stopwatch/reset",
      payload: {
        date: Date.now(),
      },
    };
    wrapper.find("button").at(1).simulate("click");
    expect(mockStore.dispatch).toHaveBeenCalledWith(action);
    expect(mockStore.dispatch).toHaveBeenCalledTimes(1);
  });

  it("should clear interval on unmount", () => {
    wrapper.unmount();
    // clearInterval is also called if stopwatch stops running, thus 2
    expect(clearInterval).toHaveBeenCalledTimes(2);
  });
});
