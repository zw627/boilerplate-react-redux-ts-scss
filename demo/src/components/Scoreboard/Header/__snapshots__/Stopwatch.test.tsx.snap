// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard/Stopwatch compare to the last snapshot 1`] = `
<Stopwatch>
  <div
    className="stopwatch"
  >
    <h2>
      Stopwatch
    </h2>
    <div
      className="stopwatch-time"
    >
      00:00.00
    </div>
    <button
      onClick={[Function]}
    >
      Start
    </button>
    <button
      onClick={[Function]}
    >
      Reset
    </button>
  </div>
</Stopwatch>
`;
